syncstorage/src/web/extractors.rs

```diff
@@ -419,49 +419,50 @@ impl FromRequest for BsoBody {
 
         let max_payload_size = state.limits.max_record_payload_bytes as usize;
 
-        let fut = <Json<BsoBody>>::from_request(req, payload)
-            .map_err(|e| {
-                warn!("⚠️ Could not parse BSO Body: {:?}", e);
-                let err: ApiError = ValidationErrorKind::FromDetails(
-                    e.to_string(),
-                    RequestErrorLocation::Body,
-                    Some("bso".to_owned()),
-                    label!("request.validate.bad_bso_body"),
-                )
-                .into();
-                err.into()
-            })
-            .and_then(move |bso: Json<BsoBody>| {
-                // Check the max payload size manually with our desired limit
-                if bso
-                    .payload
-                    .as_ref()
-                    .map(std::string::String::len)
-                    .unwrap_or_default()
-                    > max_payload_size
-                {
+        // We declare fut here to avoid moving a mutable ref of payload into the async block.
+        // See examples in actix documentation for this style of implementation.
+        let fut = <Json<BsoBody>>::from_request(req, payload);
+
+        Box::pin(async move {
+            let bso = fut
+                .map_err(|e| {
```

Comment:
This is `map_err` on the Future itself -- with it now being `.await`'d it will be more idiomatic (and simpler to reason about) to call `map_err` on the `Result` returned from the `await` itself. i.e. just change the order of calls here.

Also we can simplify this error conversion because we're now using `?` here which does the `.into()` for us. We also have a `impl From<ValidationErrorKind> for Error` so we don't need the extra `.into()` `ApiError` step either.

```
                let bso = fut
                .await
                .map_err(|e| {
                    warn!("⚠️ Could not parse BSO Body: {:?}", e);
                    ValidationErrorKind::FromDetails(
                        e.to_string(),
                        RequestErrorLocation::Body,
                        Some("bso".to_owned()),
                        label!("request.validate.bad_bso_body"),
                    )
                })?;

```